File: library-management-system/backend/src/main/java/com/library/service/UserService.java
package com.library.service;

import com.library.dto.ApiResponse;
import com.library.dto.UserDTO;
import com.library.entity.User;
import com.library.enums.Role;

import java.util.List;

public interface UserService {
    ApiResponse<List<UserDTO>> getAllUsers();
    ApiResponse<UserDTO> updateUser(Long id, UserDTO userDTO);
    ApiResponse<UserDTO> getUserByUsername(String username);
    ApiResponse<UserDTO> getUserById(Long id);
        void deleteUser(Long userId);
    User loadUserByUsername(String username);
    ApiResponse<List<UserDTO>> getUsersByRole(Role role) ;

}

File: library-management-system/frontend/src/services/userService.ts
import api from './api';
import { User, Role, PaginatedUserResponse } from '../types';

export const userService = {
  // Get all users
  getAllUsers: async (): Promise<PaginatedUserResponse<User>['data']> => {
    const response = await api.get<PaginatedUserResponse<User>>('/users');
    return response.data.data;
  },

  // Get user by ID
  getUserById: async (id: number):  Promise<PaginatedUserResponse<User>['data']> => {
    const response = await api.get<PaginatedUserResponse<User>>(`/users/${id}`);
    return response.data.data;
  },

  // Get user by username
  getUserByUsername: async (username: string):  Promise<PaginatedUserResponse<User>['data']> => {
    const response = await api.get<PaginatedUserResponse<User>>(`/users/username/${username}`);
    return response.data.data;
  },

  // Get users by role
  getUsersByRole: async (role: Role): Promise<PaginatedUserResponse<User[]>['data']> => {
    const response = await api.get<PaginatedUserResponse<User[]>>(`/users/role/${role}`);
    return response.data.data;
  },

  // Update user
  updateUser: async (id: number, user: Partial<User>):  Promise<PaginatedUserResponse<User>['data']> => {
    const response = await api.put<PaginatedUserResponse<User>>(`/users/${id}`, user);
    return response.data.data;
  },

  // Delete user
  deleteUser: async (id: number): Promise<void> => {
    await api.delete(`/users/${id}`);
  },
};
